stdlib/public/core/StringHashable.swift

```diff
@@ -12,181 +12,129 @@
+    }
   }
+}
 
-  // FIXME: cannot be marked @_versioned. See <rdar://problem/34438258>
-  // @_inlineable // FIXME(sil-serialize-all)
-  // @_versioned // FIXME(sil-serialize-all)
-  internal static func hashUTF16(
-    _ string: UnsafeBufferPointer<UInt16>,
-    into hasher: inout _Hasher
-  ) {
-    let collationIterator = _swift_stdlib_unicodeCollationIterator_create(
-      string.baseAddress!,
-      UInt32(string.count))
-    defer { _swift_stdlib_unicodeCollationIterator_delete(collationIterator) }
-
-    while true {
-      var hitEnd = false
-      let element =
-        _swift_stdlib_unicodeCollationIterator_next(collationIterator, &hitEnd)
-      if hitEnd {
-        break
+extension BidirectionalCollection where Element == UInt16, SubSequence == Self {
+  // NOT @_versioned
+  internal func hashUTF16(into hasher: inout _Hasher) {
+    var buffer = _emptyASCIIHashBuffer()
+
+    var i = startIndex
+    for cu in self {
+      defer { i = index(after: i) }
+      let cuIsASCII = cu <= 0x7F
+      let isSingleSegmentScalar = self.hasNormalizationBoundary(after: i)
+
+      guard cuIsASCII && isSingleSegmentScalar else {
+        if buffer.count != 0 {
+          hasher.append(buffer._storage)
+          buffer = _emptyASCIIHashBuffer()
+        }
+
+        let codeUnitSequence = IteratorSequence(
+          _NormalizedCodeUnitIterator(self[i..<endIndex])
+        )
+        for element in codeUnitSequence {
+          hasher.append(UInt(element))
+        }
+        return
       }
-      // Ignore zero valued collation elements. They don't participate in the
-      // ordering relation.
-      if element != 0 {
-        hasher.append(Int(truncatingIfNeeded: element))
+
+      buffer.append(UInt8(truncatingIfNeeded: cu))
+
+      if buffer.count >= buffer.capacity {
+        hasher.append(buffer._storage)
+        buffer = _emptyASCIIHashBuffer()
       }
     }
+
+    if buffer.count > 0 {
+      hasher.append(buffer._storage)
+      buffer = _emptyASCIIHashBuffer()
+    }
   }
 }
 
-#if _runtime(_ObjC)
-#if arch(i386) || arch(arm)
-    private let stringHashOffset = Int(bitPattern: 0x88dd_cc21)
-#else
-    private let stringHashOffset = Int(bitPattern: 0x429b_1266_88dd_cc21)
-#endif // arch(i386) || arch(arm)
-#endif // _runtime(_ObjC)
-
 extension _UnmanagedString where CodeUnit == UInt8 {
+  @effects(releasenone)
   @_versioned
-  @inline(never) // Hide the CF dependency
-  internal func computeHashValue() -> Int {
-#if _runtime(_ObjC)
-    let hash = _stdlib_CFStringHashCString(OpaquePointer(start), count)
-    // Mix random bits into NSString's hash so that clients don't rely on
-    // Swift.String.hashValue and NSString.hash being the same.
-    return stringHashOffset ^ hash
-#else
-    var hasher = _Hasher()
-    Unicode.hashASCII(self.buffer, into: &hasher)
-    return Int(truncatingIfNeeded: hasher.finalize())
-#endif // _runtime(_ObjC)
+  internal func hash(into hasher: inout _Hasher) {
```

Comment:
Unless you're trying to satisfy a protocol conformance, please leave the name as computeHashValue(). Otherwise this is confused with Hashable's requirement, especially if we make `hash(into:)` the name. Alternatively, UnmanagedString could be Hashable, and I'm ok with that too.